submariner-operator/broker-multi-tenancy.md

```diff
@@ -0,0 +1,321 @@
+proposes an enhancement to support multiple disjoint ClusterSet’s (e.g.,
+each belonging to a different “tenant”) hosted on a single Broker cluster.
+ The design addresses multiplicity and isolation concerns.
+
+## Motivation
+
+In SaaS offerings, such as managed OpenShift, it may be desirable to
+minimize the cost of running a dedicated Broker cluster for each ClusterSet.
+When all clusters share a high level of trust, the Broker can be co-located
+with workloads and does not require a dedicated cluster. However, this level
+of trust is neither uncommon nor desirable in a multi-tenant environment.
+Furthermore, in such settings, the Submariner Broker cluster is owned
+and managed independently of the workload clusters.
+
+Submariner repository currently has the following related issues:
+
+- [submariner#1120](https://github.com/submariner-io/submariner/issues/1120):
+Deploy multiple ClusterSet's within the same "Submariner"
+- [submariner-operator#1040](https://github.com/submariner-io/submariner-operator/issues/1040):
+Support deployment within a custom namespace
+
+In addition, a recent PR which provides partial support was recently merged:
+
+- [submariner-operator PR 1682](https://github.com/submariner-io/submariner-operator/pull/1682)
+
+### Pitfalls of the current state
+
+The current design places all resources relating to a ClusterSet in a single,
+predefined, namespace on the Broker cluster. This works well for a single
+ClusterSet and users often choose to run independent Submariner Broker clusters
+for different ClusterSets. The use of a single predefined Broker namespace
+is limiting:
+
+1. The selected namespace name should be configurable to accommodate
+requirements of different use cases.
+2. It prevents Submariner from managing multiple independent ClusterSets in a
+single Broker cluster.
+
+In a SaaS environment, we can consider three distinct user types in the system,
+with different privileges:
+
+- **Offering Admin** owns and manages the Submariner Broker(s) cluster. They are
+responsible for creating a per ClusterSet configuration space.
+- **ClusterSet/Broker Admin** manages workload clusters belonging to a ClusterSet.
+They manage cluster membership in a given ClusterSet. Admins of different ClusterSets
+should be isolated from each other.
+- **Cluster/Service Admin** manages the import and export of services from/to
+workload clusters in the ClusterSet.
+
+The first two are somewhat conflated in the current Submariner design.
+
+### Goals
+
+- Allow usage of configurable Broker namespace names.
+- Enable multiple ClusterSets to coexist in the same Submariner Broker(s)
+cluster.
+- Provide isolation (via namespaces and RBAC) between objects of different
+ClusterSets to ensure users can’t modify configuration of another user’s ClusterSet.
+- Provide clearer distinction between Offering Admin and ClusterSet Admin,
+by allowing the latter to deploy a Broker with restricted namespace permission.
```

Comment:
Currently, the broker cannot be deployed (at least by `subctl`) without having cluster-admin permissions, since it needs to create CRDs and also because it deploys the operator to a shared namespace on the broker cluster (`submariner-operator`). We would like `subctl` to be able to deploy a broker, also if the current cluster context has access to a single namespace only.